src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1314 @@
+        :nss_info: - controls how Winbind retrieves Name Service Information to construct a user's home directory and login shell. This parameter
+        is only effective if the Active Directory Domain Controller supports the Microsoft Services for Unix (SFU) LDAP schema.
+        :timeout: - timeout value for winbind-related operations. This value may need to be increased in  environments with high latencies
+        for communications with domain controllers or a large number of domain controllers. Lowering the value may cause status checks to fail.
+        :dns_timeout: - timeout value for DNS queries during the initial domain join.
+        :ldap_sasl_wrapping: - The client ldap sasl wrapping defines whether ldap traffic will be signed or signed and encrypted (sealed).
+        :createcomputer: - Active Directory Organizational Unit in which to create the NAS computer object during domain join.
+        If blank, then the default OU is used during computer account creation. Precreate the computer account in a specific OU.
+        The OU string read from top to bottom without RDNs and delimited by a "/". E.g. "createcomputer=Computers/Servers/Unix NB: A backslash
+        "\" is used as escape at multiple levels and may need to be doubled or even quadrupled. It is not used as a separator.
+        :idmap_backend: - The idmap backend provides a plugin interface for Winbind to use varying backends to store SID/uid/gid mapping tables.
+
+        The Active Directory service will be started after a configuration update if the service was initially disabled, and the updated
+        configuration sets :enable: to True. Likewise, the Active Directory service will be stopped if :enable: is changed to False
+        during an update. If the configuration is updated, but the initial :enable: state was True, then only a samba_server restart
+        command will be issued.
+        """
+        verrors = ValidationErrors()
+        old = await self.config()
+        new = old.copy()
+        new.update(data)
+        try:
+            await self.update_netbios_data(old, new)
+        except Exception as e:
+            raise ValidationError('netbiosname', str(e))
+
+        new = await self.ad_compress(new)
+
+        if not new["bindpw"] and not new["kerberos_principal"]:
+            raise ValidationError("activedirectory_update.bindname", "Bind credentials or kerberos keytab are required to join an AD domain.")
+
+        await self.middleware.call(
+            'datastore.update',
+            'directoryservice.activedirectory',
+            old['id'],
+            new,
+            {'prefix': 'ad_'}
+        )
+
+        if data['bindpw'] and data['enable'] and not old['enable']:
+            try:
+                await self.middleware.run_in_thread(self.validate_credentials)
+            except Exception as e:
+                verrors.add("activedirectory_update.bindpw", f"Failed to validate bind credentials: {e}")
```

Comment:
Raising `ValidationError` will make an impression that update was not successful when, in fact, `directoryservice.activedirectory` table row was changed. I've checked your code and have not found any obstacles to passing AD config to `self.validate_credentials` and `self.validate_domain` rather than reading it there from database directly. Am I missing something?

Reply:
No. You're not.